packages/@lwc/engine-core/src/framework/stylesheet.ts

```diff
@@ -138,16 +138,39 @@ export function getStylesheetsContent(vm: VM, template: Template): string[] {
     return content;
 }
 
+function getNearestNativeShadowComponent(vm: VM): VM | null {
+    let owner: VM | null = vm;
+    while (!isNull(owner)) {
+        if (owner.renderMode === RenderMode.Shadow && owner.shadowMode === ShadowMode.Native) {
+            return owner;
+        }
+        owner = owner.owner;
+    }
+    return owner;
+}
+
 export function createStylesheet(vm: VM, stylesheets: string[]): VNode | null {
     const { renderer, renderMode, shadowMode } = vm;
     if (renderMode === RenderMode.Shadow && shadowMode === ShadowMode.Synthetic) {
         for (let i = 0; i < stylesheets.length; i++) {
             renderer.insertGlobalStylesheet(stylesheets[i]);
         }
-        return null;
-    } else {
-        // native shadow or light DOM
+    } else if (renderer.ssr) {
```

Comment:
Side note: It will be more complex to do rehydration if SSR produces a different HTML structure than client-side rendering. @jodarove @divmain

Reply:
This is a good point, yes. And constructable styles for SSR have not been specced yet: https://github.com/WICG/webcomponents/issues/939

I consider the tradeoff worth it given the massive perf improvement in our benchmark (~30% in dom-styled-component-create-1k-same). Plus in the real-world app I tested, it was such a big difference that it was visually noticeable (~4.3s versus ~2.1s).